src/lib/structures/ty.ml

```diff
@@ -370,7 +370,12 @@ let apply_subst =
   fun s ty -> if M.is_empty s then ty else apply_subst s ty
 
 let rec fresh ty subst =
-  match ty with
+  match shorten ty with
+  | Tvar { value = Some _; _ } ->
+    (* This case is eliminated by the normalization performed
+       in [shorten]. *)
+    assert false
+
```

Comment:
`fresh` can call itself recursively on sub-terms, and `shorten` itself recurses into the term, so this introduces quadratic complexity. `shorten` should only be called once on the top-level call of `fresh` but not in the recursive calls. This can be done by shadowing `fresh`:

```ocaml
let fresh ty subst = fresh (shorten ty) subst
```